wiki/C++-Style-Guide.md

```diff
@@ -0,0 +1,814 @@
+Some of the warning are rather pedantic, but the effort required to eliminate them is worth it compared to the effort of tracking down subtle bugs that might otherwise exist.
+The "comparing signed to unsigned value" warning is particularly common, since the sizes of STL containers are unsigned, but it's common practice to use plain old "int" whenever an integer is needed, instead of using "unsigned". Assuming that your indices are never going to be larger than the maximum value of a signed int, instead of this:
+```c++
+if(index < 0 || index >= vec.size())
+```
+you can do this, which silences the warning and is also more efficient:
+```c++
+if(static_cast<size_t>(index) >= vec.size())
+```
+(If you don't understand what that statement is doing, just think of what the 2's complement binary representation of a signed number is, and what happens when that gets converted to an unsigned number.)
+
+## Avoid multiple side-effects
+
+Try not to include multiple side-effects on a single line unless it's in a context where they are expected.
+For example, consider these two for loops:
+```c++
+for(it = vec.begin(); it != vec.end(); )
+    *out++ = *it++;
+```
+```c++
+for(it = vec.begin(); it != vec.end(); ++it, ++out)
+    *out = *it;
+```
+The first option is very slightly more concise, but putting the increments into the for statement is more standard practice, and the compiler will almost certainly be able to optimize it equally well. A third, equally good option for the above would be:
+```c++
+for(it = vec.begin(); it != vec.end(); ++it)
+{
+    *out = *it;
+    ++out;
+}
+```
+This has the advantage that the "for" clauses now deal only with the "it" variable; that is, it is now a "normal" for loop.
+
+## C-style typecasts
+
+Do not use C-style typecasts.
+That is, use static_cast<>() or reinterpret_cast<>(). Try to avoid using dynamic_cast<>() or const_cast<>().
+
+## Pointers vs. references
+
+In function parameters, it may be helpful to pass constant parameters by reference, and non-constant parameters by pointer, so that it is immediately clear which parameters are being modified.
+The current code does not do this, but it seems like a good practice to move towards.
+
+## No manual memory management
+
+If at all possible, use wrapper classes instead of explicitly allocating or freeing memory.
+This makes code exception-safe, and also avoids all sorts of memory management errors. Note that std::vector always stores its contents in a contiguous block of memory, which you can access with its data() function, so there is almost never any reason to allocate your own memory instead of using vector.
+If you must use explicit memory management, always use the C++ keywords "new" and "delete", rather than the C-style malloc() and free() functions. This is because creating memory with one set of functions and deleting it with the other results in undefined behavior. Also, unlike malloc(), "new" ensures that an object is properly constructed.
```

Comment:
```suggestion
If you must use explicit memory management, always use the C++ keywords `new` and `delete`, rather than the C-style `malloc()` and `free()` functions. This is because creating memory with one set of functions and deleting it with the other results in undefined behavior. Also, unlike `malloc()`, `new` ensures that an object is properly constructed.
```